Composer/packages/client/src/pages/language-generation/code-editor.tsx

```diff
@@ -7,51 +7,79 @@ import { LgEditor, LGOption } from '@bfc/code-editor';
 import get from 'lodash/get';
 import debounce from 'lodash/debounce';
 import isEmpty from 'lodash/isEmpty';
-import { LgFile } from '@bfc/indexers';
 import { editor } from '@bfcomposer/monaco-editor/esm/vs/editor/editor.api';
 import { lgIndexer, Diagnostic } from '@bfc/indexers';
+import { RouteComponentProps } from '@reach/router';
 
 import { StoreContext } from '../../store';
 import * as lgUtil from '../../utils/lgUtil';
 
 const { check, isValid, combineMessage } = lgIndexer;
 
-interface CodeEditorProps {
-  file: LgFile;
-  template: lgUtil.Template | null;
-  line: number;
-}
-
 const lspServerPath = '/lg-language-server';
 
-export default function CodeEditor(props: CodeEditorProps) {
-  const { actions } = useContext(StoreContext);
-  const { file, template, line } = props;
+interface CodeEditorProps extends RouteComponentProps<{}> {
+  fileId?: string;
+}
+
+const CodeEditor: React.FC<CodeEditorProps> = props => {
+  const { actions, state } = useContext(StoreContext);
+  const { lgFiles } = state;
+  const file = lgFiles.length ? lgFiles[0] : null;
   const [diagnostics, setDiagnostics] = useState(get(file, 'diagnostics', []));
   const [content, setContent] = useState('');
   const [errorMsg, setErrorMsg] = useState('');
   const [lgEditor, setLgEditor] = useState<editor.IStandaloneCodeEditor | null>(null);
 
-  const fileId = file && file.id;
+  const search = props.location ? props.location.search : '';
+  const templateMatched = /^\?t=([-\w]+)/.exec(search);
+  const templateId = templateMatched ? decodeURIComponent(templateMatched[1]) : undefined;
+  const template = templateId && file ? file.templates.find(({ name }) => name === templateId) : undefined;
+
+  const hash = props.location ? props.location.hash : '';
+  const lineMatched = /L(\d+)/g.exec(hash);
+  const line = lineMatched ? +lineMatched[1] : undefined;
+
+  const fileId = (file && file.id) || 'common';
   const inlineMode = !!template;
+
   useEffect(() => {
     // reset content with file.content's initial state
     if (isEmpty(file)) return;
     const value = template ? get(template, 'body', '') : get(file, 'content', '');
     setContent(value);
-  }, [fileId, template]);
+  }, [fileId, templateId]);
 
   useEffect(() => {
-    const isInvalid = !isValid(diagnostics);
-    const text = isInvalid ? combineMessage(diagnostics) : '';
+    const currentDiagnostics =
+      inlineMode && template
+        ? diagnostics.filter(d => {
+            return (
+              d.range &&
+              d.range.start.line >= template.range.startLineNumber &&
+              d.range.end.line <= template.range.endLineNumber
+            );
+          })
+        : diagnostics;
+
+    const isInvalid = !isValid(currentDiagnostics);
+    const text = isInvalid ? combineMessage(currentDiagnostics) : '';
     setErrorMsg(text);
   }, [diagnostics]);
 
+  const editorDidMount = useCallback((lgEditor: editor.IStandaloneCodeEditor) => {
+    setLgEditor(lgEditor);
+  }, []);
```

Comment:
Any reason this is `useCallback`? If I recall, this is only called once anyways. What's the perf impact of doing this?